### PR TITLE
Add types to ConstantReader, fix issue with parts, migrate annotation helpers

### PR DESCRIPTION
**Pull request by kevmoo**

### CODE REVIEW

File: lib/src/annotation.dart [hunk 1]
```diff
@@ -2,24 +2,22 @@
...
-import 'package:path/path.dart' as p;
+
+import 'constants.dart';
+import 'type_checker.dart';
 
 dynamic instantiateAnnotation(ElementAnnotation annotation) {
   var annotationObject = annotation.constantValue;
   try {
-    return _getValue(annotationObject, annotation.element.context.typeProvider);
+    return _getValue(annotation.constantValue);
```
Comment (matanlurey) [inline lib/src/annotation.dart]: AFAIK you always want to use `computeConstantValue`.

File: lib/src/constants.dart [hunk 2]
```diff
@@ -98,6 +99,36 @@ abstract class ConstantReader {
...
+
+  /// Returns whether this constant represents a `Type` literal.
+  ///
+  /// If `true`, [typeValue] will return a `DartType` (not throw).
+  bool get isType;
+
+  /// Returns a [DartType] representing this as a `Type` value.
+  ///
+  /// Throws [FormatException] if [isType] is `false`.
+  DartType get typeValue;
```
Comment (kevmoo) [inline lib/src/constants.dart]: @matanlurey – can't get the actual `Type` because runtime vs static.

File: lib/src/annotation.dart [hunk 3]
```diff
@@ -51,33 +49,35 @@ dynamic instantiateAnnotation(ElementAnnotation annotation) {
       "${valueDeclaration.runtimeType}.");
 }
 
-dynamic _getValue(DartObject object, TypeProvider typeProvider) {
-  if (object.isNull) {
+dynamic _getValue(DartObject object) {
+  var reader = new ConstantReader(object);
+
+  if (reader.isNull) {
```
Comment (matanlurey) [inline lib/src/annotation.dart]: Optional: You can use `anyValue` for this purpose if you'd like.

File: lib/src/constants.dart [hunk 4]
```diff
@@ -180,6 +228,9 @@ class _Constant implements ConstantReader {
       _object.toBoolValue() ??
       _object.toIntValue() ??
       _object.toStringValue() ??
+      _object.toDoubleValue() ??
+      (isSymbol ? this.symbolValue : null) ??
```
Comment (kevmoo) [inline lib/src/constants.dart]: Special case here to return an actual `Symbol` and not a String

File: lib/src/constants.dart [hunk 5]
```diff
@@ -221,6 +272,28 @@ class _Constant implements ConstantReader {
...
+  Symbol get symbolValue => isSymbol
+      ? new Symbol(_object.toSymbolValue())
+      : _throw('Symbol', _object);
+
+  @override
+  bool get isType => _object.toTypeValue() != null;
+
+  @override
+  DartType get typeValue =>
+      isType ? _object.toTypeValue() : _throw("Type", _object);
```
Comment (matanlurey) [inline lib/src/constants.dart]: Consistent use of ' please.

File: lib/src/constants.dart [hunk 6]
```diff
@@ -160,6 +190,24 @@ class _NullConstant implements ConstantReader {
   bool get isString => false;
 
   @override
+  bool get isDouble => false;
+
+  @override
+  double get doubleValue => _throw("double");
```
Comment (matanlurey) [inline lib/src/constants.dart]: '

File: lib/src/constants.dart [hunk 7]
```diff
@@ -160,6 +190,24 @@ class _NullConstant implements ConstantReader {
...
+
+  @override
+  bool get isSymbol => false;
+
+  @override
+  Symbol get symbolValue => _throw("Symbol");
```
Comment (matanlurey) [inline lib/src/constants.dart]: '

File: lib/src/constants.dart [hunk 8]
```diff
@@ -160,6 +190,24 @@ class _NullConstant implements ConstantReader {
...
+
+  @override
+  get isType => false;
+
+  @override
+  DartType get typeValue => _throw("Type");
```
Comment (matanlurey) [inline lib/src/constants.dart]: '

File: lib/src/constants.dart (hunk 6 above)
Comment (kevmoo) [inline lib/src/constants.dart]: Huh?

Comment (kevmoo) [inline lib/src/constants.dart]: Oh...geez...

File: lib/src/annotation.dart (hunk 3 above)
Comment (kevmoo) [inline lib/src/annotation.dart]: Sadly no, we need to talk about that.

Type, List, and Map are weird – you get DartType.

I think we should update `anyValue` to exclude those and explain why.

File: lib/src/constants.dart (hunk 6 above)
Comment (kevmoo) [inline lib/src/constants.dart]: done

File: lib/src/constants.dart (hunk 7 above)
Comment (kevmoo) [inline lib/src/constants.dart]: done

File: lib/src/constants.dart (hunk 8 above)
Comment (kevmoo) [inline lib/src/constants.dart]: done

File: lib/src/constants.dart (hunk 5 above)
Comment (kevmoo) [inline lib/src/constants.dart]: done – enable a lint 😁

File: lib/src/annotation.dart (hunk 1 above)
Comment (kevmoo) [inline lib/src/annotation.dart]: Just ran a little test – in all cases `constantValue` and `computeConstantValue` return the same result. 🤷‍♂️

**Pull request merged**